### ISSUE TITLE
Convert velocities between Topocentric, Geocentric, and Heliocentric frames

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#3544]
It would be nice to have functionality to convert between all these frames of reference. It seems we have all the pieces of the puzzle, including `SkyCoord`, `Time`, `EarthLocation`, and access to all wrapped ERFA functions (unless I'm missing something?).

EDIT: this would be a good GSoC project, and the people to contact are @eteq and @adrn

**Comment by astrofrog** [comment]
I feel like there is going to start being some overlap with SunPy - in particular for the ``Heliocentric`` frame.

**Comment by sahmed95** [comment]
Isn't it listed as a GSoc project ?

**Comment by astrofrog** [comment]
@sahmed95 - indeed it is, just noticed it! It would indeed be a nice GSoC project.

**Comment by sahmed95** [comment]
Yeah, will be a nice addition to SkyCoord I guess

**Comment by sahmed95** [comment]
There are code snippets here and there to do velocity transformations. Like this one written by some Nasa guy..... "http://idlastro.gsfc.nasa.gov/ftp/pro/astro/glactc_pm.pro " 
But no package has a unified library for that yet. I am interested in implementing this.

**Comment by mhvk** [comment]
See also #2244 - and, yes, we should have this! Including of course barycentric velocities.

**Comment by eteq** [comment]
Note that this is a major undertaking because we have to build an entire framework from scratch for velocities to do this, which we have to get right due to the fact that it's related to PMs, RV measurements, which are high-precision bread-and-butter sort of astronomy.  @adrn and I have talked about this as the next major `coordinates` feature, but it definitely requires some careful planning.

**Comment by sahmed95** [comment]
Is this a good start for calculation of galactic space velocities ?

http://articles.adsabs.harvard.edu/cgi-bin/nph-iarticle_query?bibcode=1987AJ.....93..864J&db_key=AST&page_ind=0&data_type=GIF&type=SCREEN_VIEW&classic=YES

I wish to take this up as a GSoC project and need some pointers as I intend to include a simple code implementing basic transformation, for my application. I am no package expert, but I have been looking into the coordinates package and I have some idea of whats happening under the hood. 
Mentors ?

**Comment by keflavich** [comment]
:+1:, also should include LSR conversions

**Comment by Juanlu001** [comment]
> And we have to get that really right due to the fact that it's critical for PMs, RV measurements, and similar stuff that is high-precision bread-and-butter sort of astronomy.

And also for astrodynamics :) Once velocity measurements land into astropy, I will be able to use them in [poliastro](https://github.com/poliastro/poliastro) and take advantage of the frame conversion routines.

**Comment by mhvk** [comment]
Just a link back to some discussion in https://github.com/astropy/specutils/pull/112, and particularly  https://gist.github.com/StuartLittlefair/5aaf476c5d7b52d20aa9544cfaa936a1

The main question now is where to put this. `Time` already has `light_travel_time`, so it might be OK to put a barycentric velocity there as well. On the other hand, perhaps most people would try to find it in a `SkyCoord`. Maybe, in fact, this is one example where it makes sense to have both? Both methods could call a common function, just swapping the arguments around a bit.

**Comment by eteq** [comment]
To add more to https://github.com/astropy/astropy/issues/3544#issuecomment-256345519 - this connects back to the bigger question of whether velocity frames *in general* should be a thing tied to coordinate frames, or independent, or loosely coupled.

**Comment by eteq** [comment]
In *principal* this was addressed by the velocities framework now in v2.0.  I'm not sure we've *tested* all these use cases, but that's a subtly different question so I think this issue can now be closed.
